relay-otel/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+//! OpenTelemetry to Sentry transformation utilities.
+//!
+//! This crate provides common functionality for converting OpenTelemetry data structures
+//! to Sentry format. It serves as a shared library for both span and log transformations.
+
+#![warn(missing_docs)]
+#![doc(
+    html_logo_url = "https://raw.githubusercontent.com/getsentry/relay/master/artwork/relay-icon.png",
+    html_favicon_url = "https://raw.githubusercontent.com/getsentry/relay/master/artwork/relay-icon.png"
+)]
+
+use opentelemetry_proto::tonic::common::v1::any_value::Value as OtelValue;
+use relay_event_schema::protocol::{Attribute, AttributeType};
+use relay_protocol::Value;
+
+/// Converts an OpenTelemetry AnyValue to a Sentry attribute.
+///
+/// This function handles the conversion of OpenTelemetry attribute values to Sentry attribute types.
+/// Complex types like arrays and key-value lists are serialized to JSON strings for safety and
+/// compatibility.
+///
+/// For array and key-value list values, this function filters out nested complex types
+/// (nested arrays and key-value lists) before serialization to prevent issues with
+/// the OTLP protocol and ensure safe handling.
+pub fn otel_value_to_attribute(otel_value: OtelValue) -> Option<Attribute> {
+    let (ty, value) = match otel_value {
+        OtelValue::StringValue(s) => (AttributeType::String, Value::String(s)),
+        OtelValue::BoolValue(b) => (AttributeType::Boolean, Value::Bool(b)),
+        OtelValue::IntValue(i) => (AttributeType::Integer, Value::I64(i)),
+        OtelValue::DoubleValue(d) => (AttributeType::Double, Value::F64(d)),
+        OtelValue::BytesValue(bytes) => {
+            let s = String::from_utf8(bytes).ok()?;
+            (AttributeType::String, Value::String(s))
+        }
+        OtelValue::ArrayValue(array) => {
+            // Filter out nested arrays and key-value lists for safety.
+            // This is not usually allowed by the OTLP protocol, but we filter
+            // these values out before serializing for robustness.
```

Comment:
What's the reason we do not skip these values like we did before?

Reply:
We always had the `filter_map` filtering, are you referring to the added `KvlistValue` logic?

`KvlistValue` value attributes exist on logs, but not spans so this branch needed to be added.